### PR TITLE
Upgrade sqlparser-rs to 0.51.0, support new interval logic from `sqlparse-rs`

### LINKED ISSUE #12190: Incorrect behavior of arithmetic operations between time values
### Describe the bug

Consider the following query:

```
DataFusion CLI v41.0.0
> select ('2024-08-27T08:21:27Z'::timestamp + interval '1 day' -  interval '1 second' - interval '1 day');
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Utf8("2024-08-27T08:21:27Z") + IntervalMonthDayNano("IntervalMonthDayNano { months: 0, days: 1, nanoseconds: 0 }") - IntervalMonthDayNano("IntervalMonthDayNano { months: 0, days: 0, nanoseconds: 1000000000 }") - IntervalMonthDayNano("IntervalMonthDayNano { months: 0, days: 1, nanoseconds: 0 }") |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 2024-08-29T08:21:26                                                                                                                                                                                                                                                                                     |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row(s) fetched.
Elapsed 0.005 seconds.
```

While the expected result should be `2024-08-27 08:21:26.000000` .

When I add parentheses to the query, the results are what I expect:
```
> select ('2024-08-27T08:21:27Z'::timestamp + interval '1 day' -  interval '1 second') - interval '1 day';
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Utf8("2024-08-27T08:21:27Z") + IntervalMonthDayNano("IntervalMonthDayNano { months: 0, days: 1, nanoseconds: 0 }") - IntervalMonthDayNano("IntervalMonthDayNano { months: 0, days: 0, nanoseconds: 1000000000 }") - IntervalMonthDayNano("IntervalMonthDayNano { months: 0, days: 1, nanoseconds: 0 }") |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 2024-08-27T08:21:26                                                                                                                                                                                                                                                                                     |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row(s) fetched.
Elapsed 0.002 seconds.
```

The issue seems to happening when you have multiple (more than 2?) operands in arithmetic. e.g.:
```
> select ('2024-08-27T08:21:27Z'::timestamp + interval '1 day' -  interval '1 day' - interval '2 day');
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Utf8("2024-08-27T08:21:27Z") + IntervalMonthDayNano("IntervalMonthDayNano { months: 0, days: 1, nanoseconds: 0 }") - IntervalMonthDayNano("IntervalMonthDayNano { months: 0, days: 1, nanoseconds: 0 }") - IntervalMonthDayNano("IntervalMonthDayNano { months: 0, days: 2, nanoseconds: 0 }") |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 2024-08-29T08:21:27                          -- Should be     2024-08-25 08:21:27.000000                                                                                                                                                                                                                              |
+------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row(s) fetched.
Elapsed 0.001 seconds.
```

When you have two operands, the results seem to match what you'd expect:
```
> select ('2024-08-27T08:21:27Z'::timestamp + interval '1 day' - interval '2 day');
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Utf8("2024-08-27T08:21:27Z") + IntervalMonthDayNano("IntervalMonthDayNano { months: 0, days: 1, nanoseconds: 0 }") - IntervalMonthDayNano("IntervalMonthDayNano { months: 0, days: 2, nanoseconds: 0 }") |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 2024-08-26T08:21:27                                                                                                                                                                                      |
+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row(s) fetched.
Elapsed 0.002 seconds.
```


### To Reproduce

You can use the same queries mentioned above.

### Expected behavior

Mentioned above as well

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by samuelcolvin**
## Which issue does this PR close?

~WIP adopt https://github.com/sqlparser-rs/sqlparser-rs/pull/1398~

fixes #12190

## Rationale for this change

Interval parsing is a bit of a mess, this is fixed I think in https://github.com/sqlparser-rs/sqlparser-rs/pull/1398, this adopts that change and removes some hacks

## What changes are included in this PR?

* ~use branch from https://github.com/sqlparser-rs/sqlparser-rs/pull/1398~
* Update to latest sqlparser-rs release
* fix `sql_interval_to_expr`
* fix tests

## Are these changes tested?

Yes

## Are there any user-facing changes?

Not AFAIK.

### CODE REVIEW

File: Cargo.toml
Comment (findepi) [inline Cargo.toml:138]: can you please update to 0.51.0 now that it's released?

**Comment by findepi** [comment]
nit: We could have "fixes https://github.com/apache/datafusion/issues/12190" in the PR description to let the issue auto-close when PR is merged.

**Comment by samuelcolvin** [comment]
@alamb I believe this is ready to review/merge.

File: datafusion/sql/src/expr/value.rs
Comment (alamb) [inline datafusion/sql/src/expr/value.rs:199]: It is great to remove all this stuff from the sql planner (as it is now in the parser)

File: datafusion/sql/tests/cases/plan_to_sql.rs
Comment (alamb) [inline datafusion/sql/tests/cases/plan_to_sql.rs:498]: this is a driveby (unrelated) cleanup, right? (this is fine I am just verifying my understanding)

File: datafusion/sqllogictest/test_files/expr.slt
Comment (alamb) [inline datafusion/sqllogictest/test_files/expr.slt:1394]: 🎉

File: datafusion/sqllogictest/test_files/interval_mysql.slt
Comment (alamb) [inline datafusion/sqllogictest/test_files/interval_mysql.slt:21]: I believe technically this is a regression in that this query used to work. 

However, as this PR shows, it works in a different dialect

I think it would be possible to add coercion rules to automatically coerce the `Utf8` to Interval to make this query continue work. I can file a follow on ticket

Comment (alamb) [reply datafusion/sqllogictest/test_files/interval_mysql.slt:21]: I double checked and this works today in datafusion-cli so this is technically a regression in functionality

```shell
andrewlamb@Andrews-MacBook-Pro-2:~/Software/influxdb_iox$ datafusion-cli
DataFusion CLI v41.0.0
> select interval '1' + '1' month;
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| IntervalMonthDayNano("IntervalMonthDayNano { months: 1, days: 0, nanoseconds: 0 }") + IntervalMonthDayNano("IntervalMonthDayNano { months: 1, days: 0, nanoseconds: 0 }") |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| 2 mons                                                                                                                                                                    |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row(s) fetched.
Elapsed 0.008 seconds.
```

**Review by alamb — APPROVED** [review]
Thank you @samuelcolvin  -- this looks great to me. ❤️ 

I think we should merge this after the 42.0.0 release is cut (scheduled for today) to give it "maximum bake time" on main (aka so we can test with early adopters before releasing to crates.io)

I think technically this PR also introduces a regression/change in parsing certain interval arithmetic (see comments inline) but I also think we can support the syntax again with some changing of the coercion rules. 

Let's file a follow on ticket to track that work before we merge this pr (I can do so if no one beats me to it)

**Comment by alamb** [comment]
Thank you also @findepi  for the review

File: datafusion/sql/tests/cases/plan_to_sql.rs
Comment (samuelcolvin) [reply datafusion/sql/tests/cases/plan_to_sql.rs:498]: It was necessary to get tests to pass, I assume it came from some other change in sqlparser, I didn't look too hard.

But it's not just cosmetic.

File: datafusion/sqllogictest/test_files/interval_mysql.slt
Comment (samuelcolvin) [reply datafusion/sqllogictest/test_files/interval_mysql.slt:21]: well it "works" but it leads to many very confusing scenarios and doesn't match the behaviour of ANSI SQL or any other SQL dialect.

It's less permissive but also less ambiguos, I wouldn't call it a regression.

File: datafusion/sql/tests/cases/plan_to_sql.rs
Comment (findepi) [reply datafusion/sql/tests/cases/plan_to_sql.rs:498]: this mostly changes `catalog` to `"catalog"`
a driveby is usage or raw literals (which is a nice thing on its own)

File: datafusion/sqllogictest/test_files/interval_mysql.slt
Comment (alamb) [reply datafusion/sqllogictest/test_files/interval_mysql.slt:21]: makes sense. There is a workaround as well (use the mysql dialect).

**Comment by alamb** [comment]
The 42.0.0 release has been cut -- let's start the code flowing!

File: datafusion/functions/src/datetime/date_part.rs
Comment (samuelcolvin) [inline datafusion/functions/src/datetime/date_part.rs:189]: this should mean epoch is now supported.

**Pull request merged**
